File: Dynamic Programming/Palindrome.java
import java.util.*;

public class Palindrome {


  public boolean isPalindrome(String s) {
    int len = s.length();

    for (int i=0; i < len/2; i++) {
        if (s.charAt(i) != s.charAt(len-1-i)) return false;
    }
    return true;
  }

  public boolean isAnyPermutationAPalindrome(String s) {

    int len = s.length();
    Map<Character, Integer> charFreq = new HashMap<>(len);

    for (char c : s.toCharArray()) {

        if (charFreq.containsKey(c)) {
          charFreq.put(c, charFreq.get(c) + 1);
        } else {
          charFreq.put(c, 1);
        }

    }

    int oddCharFreq = 0;
    for (Integer x : charFreq.values()) {
        oddCharFreq += (x % 2 == 0) ? 0 : 1;
    }

    if (len % 2 == 0) {
      return oddCharFreq == 0;
    } else {
      return oddCharFreq == 1;
    }

  }

  // Dynamic Programming
  // Store solutions to subproblems
  public String longestPalindromeInString(String s) {

    int len = s.length();
    Boolean[][] isPalindromTable = new Boolean[len][len];

    // A single character is a palindrome of length one
    for (int i = 0; i < len; i++) {
        isPalindromTable[i][i] = true;
    }
    int maxLength = 1;
    int startingIndex = 0;

    // Check two character plaindromes
    for (int i = 0; i < len-1; i++) {

      if (s.charAt(i) == s.charAt(i+1)) {
        isPalindromTable[i][i+1] = true;
        maxLength = 2;
        startingIndex = i;
      }

    }

    // Searching for palindromes of length 3 and greater (up until the length of the input string)
    for (int subLen = 3; subLen <= len; subLen++) {

      // The starting index of the substring in question
      for (int i=0; i <= len - subLen; i++) {

          int e = i + subLen - 1;
          if  (isPalindromTable[i+1][e-1] && s.charAt(i) == s.charAt(e) ) {
              isPalindromTable[i][e] = true;
              startingIndex = i;
              maxLength = subLen;
          }

      }

    }

    return s.substring(startingIndex , startingIndex + maxLength);

  }


  public int longestPalindromicSubsequence(String s) {

    int len = s.length();
    int[][] subproblems = new int [len][len];

    for (int i = 0; i < len; i++) {
      subproblems[i][i] = 1;
    }

    for (int i = 0; i < len - 1; i++) {

      if (s.charAt(i) == s.charAt(i+1)) {
        subproblems[i][i+1] = 2;
      } else {
        subproblems[i][i+1] = 1;
      }

    }

    for (int k = 3; k <= len; k++) {
      for (int i = 0; i <= len - k; i++) {
        int j = i + k - 1;
        if (s.charAt(i) == s.charAt(j)) {
          subproblems[i][j] = subproblems[i+1][j-1] + 2;
        } else {
          subproblems[i][j] = Math.max(subproblems[i+1][j] , subproblems[i][j-1]);
        }
      }

    }

    return subproblems[0][len-1];

  }



}

File: Dynamic Programming/Palindrome.js
/*jslint es6 */
'use strict';

function isPalindrome(word) {
    word = word.toLowerCase();
    for (let i=0; i < word.length/2; i++) {
        if (word.charAt(i) != word.charAt(word.length - 1 -i)) return false;
    }
    return true;
}


function isAnyPermutationAPalindrome(word) {
    word = word.toLowerCase();
    const charFreq = new Set();

    for (const char of word) {
        if (charFreq.has(char)) {
            charFreq.delete(char);
        } else {
            charFreq.add(char);
        }
    }

    return charFreq.size > 1 ? false : true;

}


function longestPalindromeInString(word) {
    word = word.toLowerCase();

    const isPalindromTable = new Array(word.length);
    for (let i=0; i < word.length; i++) {
        isPalindromTable[i] = new Array(word.length).fill(false);
    }


    for (let i=0; i < word.length; i++) {
        isPalindromTable[i][i] = true;
    }

    let maxLength = 1;
    let startingIndex = 0;

    for (let i=0; i < word.length -1; i++) {
        if(word.charAt(i) == word.charAt(i+1)) {
            isPalindromTable[i][i+1] = true;
            maxLength = 2;
            startingIndex = i;
        }
    }

    for (let sublen=3; sublen <= word.length; sublen++ ) {
        for (let i=0; i <= word.length - sublen; i++) {
            if (word.charAt(i) == word.charAt(i+sublen-1) && isPalindromTable[i+1][i+sublen-2]) {
                isPalindromTable[i][i+sublen-1] = true;
                maxLength = sublen;
                startingIndex = i;
            }
        }
    }

    return word.substring(startingIndex, startingIndex + maxLength);
}

function longestPalindromicSubsequence(word) {
    word = word.toLowerCase();
    const len = word.length;

    const palindromicSubsequenceTable = new Array(len);
    for (let i=0; i < len; i++) {
        palindromicSubsequenceTable[i] = new Array(len).fill(1);
    }

    for (let i = 0; i < len -1; i++) {
        if (word.charAt(i) == word.charAt(i+1)) palindromicSubsequenceTable[i][i+1] = 2;
    }

    for (let subProblem = 3; subProblem <= len; subProblem++) {
        for (let i = 0; i <= len - subProblem; i++) {
            if (word.charAt(i) == word.charAt(i+subProblem -1)) {
                palindromicSubsequenceTable[i][i+subProblem -1] = palindromicSubsequenceTable[i+1][i+subProblem -2] + 2;
            } else {
                palindromicSubsequenceTable[i][i+subProblem -1] = Math.max(palindromicSubsequenceTable[i+1][i+subProblem -1], palindromicSubsequenceTable[i][i+subProblem -2]);
            }
        }
    }
    return palindromicSubsequenceTable[0][len-1];
}

console.log(isPalindrome("Haider"));
console.log(isAnyPermutationAPalindrome("bccBdz"));
console.log(longestPalindromeInString("ZbccBdz"));
console.log(longestPalindromicSubsequence("xybccBdz"));

module.exports = {
    isPalindrome,
    isAnyPermutationAPalindrome,
    longestPalindromeInString
};
